File: toys/cbox/compiler/gcc_plugins/tu_parse/build_adp_parser.pl
#!/usr/bin/perl

use Inline::Files;

my @pair;
my @full_names;

while (<ADP_PARSER_BEGIN>) {
    print $_;
}

while (<>) {
    chomp;
    #print $_;
    undef @pair;
    undef @full_names;
    @pair = split(/\,/, $_);
    #print @pair;
    $name = @pair[0];
    $type = @pair[1];
    my @full_names = split(/\./, $name);
    #print "$name, $type\n";
    if ($type =~ /BOOLEAN_Type/) {
        $_ =~ /BOOLEAN_Type, (\d+)/;
        $shift = $1;
        print "    parser->Set_Column_Masked(++column, \"$full_names[1]\", TELEMETRY_LOG_STRUCT_ADP, u.$full_names[0], $type, (1 << $shift));\n"
    #elsif (length(@full_names > 1)) {
    #    print "parser->Set_Column(++column, \"$full_names[1]\", TELEMETRY_LOG_STRUCT, $name, $type)\n"
    } else {
        print "    parser->Set_Column(++column, \"$name\", TELEMETRY_LOG_STRUCT_ADP, $name, $type);\n"
    }
}
print "\n";

while (<ADP_PARSER_END>) {
    print $_;
}

__ADP_PARSER_BEGIN__
static Parser_Alias * Create_ADP_Telemetry_Parser(const std::string & filename)
{    
    int column = -1;
    Parser<TELEMETRY_LOG_STRUCT_ADP> * parser = new Parser<TELEMETRY_LOG_STRUCT_ADP>(filename);
    parser->Duplicate_To_Combined_File(true);

    parser->Set_Lines_Per_File(25000);


__ADP_PARSER_END__
    return parser;
}

File: toys/misc/py/Telem/build_adp_parser.py
#!/usr/bin/env python3

import re
import sys
import ap_strs

re_bool_type= "BOOLEAN_Type :(\d+)"
re_flag_type= "flag"
period = "\."

print(ap_strs.ADP_PARSER_BEGIN)

try:
    for line in sys.stdin:
        if line:
            line = line.rstrip()
            pair = line.split(',', 2)
            name = pair[0];
            type = pair[1];
            boolval = re.search(re_bool_type, line)
            flagval = re.search(re_flag_type, line)
            if (flagval):
                shift = boolval.group(1)
                full_names = name.split('.')
                print("    parser->Set_Column_Masked(++column, \"{0}\", TELEMETRY_LOG_STRUCT_ADP, {1}.{2}, BOOLEAN_Type, (1 << {3}));".format(full_names[2], full_names[0], full_names[1], shift))
            elif (boolval):
                shift = boolval.group(1)
                full_names = name.split('.')
                print("    parser->Set_Column_Masked(++column, \"{0}\", TELEMETRY_LOG_STRUCT_ADP, {1}.{2}, BOOLEAN_Type, (1 << {3}));".format(full_names[2], full_names[0], full_names[1], shift))
            else:
                print("    parser->Set_Column(++column, \"{0}\", TELEMETRY_LOG_STRUCT_ADP, {1}, {2});".format(name, name, type))
                pass

except Exception:
    print("error in data")
finally:
    pass

print(ap_strs.ADP_PARSER_END)


